### ISSUE TITLE
no-use-before-define with global variable reassign

### ISSUE THREAD

**Issue by Mantish** [eslint/eslint#10767]
Hi,
`no-use-before-define` triggers an error for this kind of code:

`var global = global || 'default';`

The variable is defined in the `globals` section of the eslint file. So in my opinion, it shouldn't trigger the error.

What do you think?

**My environment**

* **ESLint Version: 4.2.0**
* **Node Version: 8.9.1**
* **npm Version: 5.5.1**
* **default parser**

**Comment by platinumazure** [comment]
Hi @Mantish, thanks for the issue.

It looks like the rule is triggering because you're redeclaring the variable. If you remove the `var` keyword and just use/reassign the variable without a new declaration, then the rule won't report. (See [demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZ2xvYmFsIGZvbywgYmFyICovXHJcbi8qIGVzbGludCBuby11c2UtYmVmb3JlLWRlZmluZTogZXJyb3IgKi9cclxuXHJcbnZhciBmb28gPSBmb28gfHwgXCJkZWZhdWx0XCI7XHJcblxyXG5iYXIgPSBiYXIgfHwgXCJkZWZhdWx0XCI7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19) for an example.)

**Comment by Mantish** [comment]
Hi @platinumazure,
Thanks for the reply and the demo.

It's true that in only happens when the variable is (re)declared, but I'd question if that's the intended behavior.
To my understanding, this rule should never report for variables defined as globals.

**Comment by platinumazure** [comment]
@Mantish I think it is intended because with the redeclare, you're effectively working with two separate declarations. One (invisible/outside the file) declaration for the global, and another for the `var` statement.

Instead of redeclaring the variable, it should just be used/reassigned.

**Comment by Mantish** [comment]
@platinumazure if that is the case I find it very confusing.
Wouldn't that belong to other kind of rules, like `no-global-assign` or `no-redeclare`?

**Comment by kaicataldo** [comment]
I think this is actually the correct behavior. When it's being redeclared, `global` is actually undefined because it is now referencing the variable that was just declared. When it's reassigned, it references the previously declared variable in the outer/global scope. Check out this [quick illustration](http://jsbin.com/tebuyacava/edit?js,console) I put together in JS Bin.

**Comment by Mantish** [comment]
Hi @kaicataldo thanks for the jsbin example, it's very useful.
However, there's a difference when doing the re-declaration in the global scope.
In your example the variable is re-declared inside the function scope, which makes it undefined.
When working in the global scope, re-declaring the variable doesn't make it loose it's value.

I realize this is probably an edge-case but it would be nice if considered when applying the rule.

**Comment by brendanarnold** [comment]
Similary I found the following causes the same lint rule to be triggered ...

```
const { services } = require('./my-services')  // Lint error here ...

const listServices = (a = services.a) => { // ... and lint error here
  const services = [] // Same variable name used in a different context
}

listServices()
```

Removing or changing the local definition of `services` cleared the linting errors
